### PR TITLE
[new release] multicont (1.0.2)

### PR DESCRIPTION
**Pull request by dhil**
Multi-shot continuations in OCaml

- Project page: <a href="https://github.com/dhil/ocaml-multicont">https://github.com/dhil/ocaml-multicont</a>

##### CHANGES:

This release adds support for the anticipated release of OCaml 5.2.

Changes:

* Patch dhil/ocaml-multicont#7: OCaml 5.2 support (thanks to kit-ty-kate for the issue
  report dhil/ocaml-multicont#6; thanks to David Allsopp for reviewing the patch).  The
  change accounts for the new continuation representation.
* Added a basic testsuite runnable via `dune runtest`.
* Fixed a memory leak in the rollback parsing example.
* Added an entry about subtle interactions of unrestricted and linear
  effects in the "Cautionary tales" section of the README.

### CODE REVIEW

File: packages/multicont/multicont.1.0.2/opam [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+maintainer: ["Daniel Hillerström"]
+authors: ["Daniel Hillerström"]
+license: "MIT"
+tags: ["multi-shot continuations" "effect handlers"]
+homepage: "https://github.com/dhil/ocaml-multicont"
+bug-reports: "https://github.com/dhil/ocaml-multicont/issues"
+depends: [
+  "ocaml" {":" >= "5.0.0"}
+  "dune" {>= "3.8"}
+  "dune-configurator" {":" >= "3.8"}
```
Comment (kit-ty-kate) [inline packages/multicont/multicont.1.0.2/opam:14]:
```suggestion
  "ocaml" {>= "5.0.0"}
  "dune" {>= "3.8"}
  "dune-configurator" {>= "3.8"}
```

**Comment by kit-ty-kate** [comment]
Note for opam-repository maintainers: this package adds 5.2 support and any previous version should be constrain on `< 5.2` (segfaults at runtime)

**Comment by mseri** [comment]
Looks like it fails on bytecode only:
```
#=== ERROR while compiling multicont.1.0.2 ====================================#
# context              2.2.0~beta2~dev | linux/arm32 | ocaml-base-compiler.5.1.1 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.2/multicont-1.0.2.tbz)
# path                 ~/.opam/5.1/.opam-switch/build/multicont.1.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 79 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/multicont-7-b04c7b.env
# output-file          ~/.opam/log/multicont-7-b04c7b.out
### output ###
# File "test/dune", line 4, characters 8-21:
# 4 |         generic_count
#             ^^^^^^^^^^^^^
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc -w -40 -g -o test/generic_count.exe -output-complete-exe lib/multicont.cma /home/opam/.opam/5.1/lib/ocaml/unix/unix.cma /home/opam/.opam/5.1/lib/stdlib-shims/stdlib_shims.cma /home/opam/.opam/5.1/lib/ounit2/advanced/oUnitAdvanced.cma /home/opam/.opam/5.1/lib/ounit2/oUnit.cma test/.async.eobjs/byte/dune__exe.cmo test/.async.eobjs/byte/dune__exe__Generic_count.cmo)
# /usr/bin/ld: cannot find -lmulticont_stubs_byte: No such file or directory
# collect2: error: ld returned 1 exit status
# File "_none_", line 1:
# Error: Error while building custom runtime system
# File "test/dune", line 3, characters 8-14:
# 3 |         choice
#             ^^^^^^
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc -w -40 -g -o test/choice.exe -output-complete-exe lib/multicont.cma /home/opam/.opam/5.1/lib/ocaml/unix/unix.cma /home/opam/.opam/5.1/lib/stdlib-shims/stdlib_shims.cma /home/opam/.opam/5.1/lib/ounit2/advanced/oUnitAdvanced.cma /home/opam/.opam/5.1/lib/ounit2/oUnit.cma test/.async.eobjs/byte/dune__exe.cmo test/.async.eobjs/byte/dune__exe__Choice.cmo)
# /usr/bin/ld: cannot find -lmulticont_stubs_byte: No such file or directory
# collect2: error: ld returned 1 exit status
# File "_none_", line 1:
# Error: Error while building custom runtime system
# File "test/dune", line 2, characters 8-13:
# 2 |  (names async
#             ^^^^^
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc -w -40 -g -o test/async.exe -output-complete-exe lib/multicont.cma /home/opam/.opam/5.1/lib/ocaml/unix/unix.cma /home/opam/.opam/5.1/lib/stdlib-shims/stdlib_shims.cma /home/opam/.opam/5.1/lib/ounit2/advanced/oUnitAdvanced.cma /home/opam/.opam/5.1/lib/ounit2/oUnit.cma test/.async.eobjs/byte/dune__exe.cmo test/.async.eobjs/byte/dune__exe__Async.cmo)
# /usr/bin/ld: cannot find -lmulticont_stubs_byte: No such file or directory
# collect2: error: ld returned 1 exit status
# File "_none_", line 1:
# Error: Error while building custom runtime system
# File "test/dune", line 5, characters 8-15:
# 5 |         nqueens)
#             ^^^^^^^
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc -w -40 -g -o test/nqueens.exe -output-complete-exe lib/multicont.cma /home/opam/.opam/5.1/lib/ocaml/unix/unix.cma /home/opam/.opam/5.1/lib/stdlib-shims/stdlib_shims.cma /home/opam/.opam/5.1/lib/ounit2/advanced/oUnitAdvanced.cma /home/opam/.opam/5.1/lib/ounit2/oUnit.cma test/.async.eobjs/byte/dune__exe.cmo test/.async.eobjs/byte/dune__exe__Nqueens.cmo)
# /usr/bin/ld: cannot find -lmulticont_stubs_byte: No such file or directory
# collect2: error: ld returned 1 exit status
# File "_none_", line 1:
# Error: Error while building custom runtime system
```

**Comment by dhil** [comment]
Seems to be all failures with linking the tests against the bytecode library. I am not sure why that would happen. My test stanza looks like
```dune
(tests
 (names async
        choice
        generic_count
        nqueens)
 (libraries multicont ounit2))
```
Do I need to add something specific for bytecode only compilation?

**Comment by kit-ty-kate** [comment]
It could be due to the fact that you have two different `(foreign_stubs ...)` (one for byte, one for native) when defining the multicont library. Maybe dune doesn't like that for some reason. It might be useful to create an issue in the dune bugtracker, maybe the maintainers have some inputs on that.

**Comment by dhil** [comment]
> It could be due to the fact that you have two different `(foreign_stubs ...)` (one for byte, one for native) when defining the multicont library. Maybe dune doesn't like that for some reason. It might be useful to create an issue in the dune bugtracker, maybe the maintainers have some inputs on that.

Looks like a similar issue has already been reported https://github.com/ocaml/dune/issues/6073.

**Comment by dhil** [comment]
I have implemented a variation of the workaround discussed in the linked issue. It seems to have made the CI happy.

**Comment by mseri** [comment]
Thanks

**Pull request merged**